api/src/auth_routes.ts

```diff
@@ -77,69 +138,164 @@ export function add_auth_routes(app: any, handlers: any) {
-      // user already registered, sign them up for this notebook
-      // should there be conditions on this? Eg. check the email.
-      await acceptInvite(req.user, invite);
-      req.flash(
-        'message',
-        'You will now have access to the ${invite.notebook} notebook.'
-      );
-      res.redirect('/');
-    } else {
-      // need to sign up the user, show the registration page
-      const available_provider_info = [];
-      for (const handler of CONDUCTOR_AUTH_PROVIDERS) {
-        available_provider_info.push({
-          label: handler,
-          name: AVAILABLE_AUTH_PROVIDER_DISPLAY_INFO[handler].name,
+
+    // Generate a token
+    const token = await generateUserToken(user);
+
+    // Append the token to the redirect URL
+    const redirectUrlWithToken = `${redirect}?token=${token.token}`;
+
+    // Redirect to the app with the token
+    return res.redirect(redirectUrlWithToken);
+  };
+
+  /**
+   * Handle local login request with username and password
+   */
+  app.post<{}, {}, {}, RequestQueryRedirect>(
+    '/auth/local',
+    (req, res, next: NextFunction) => {
+      const redirect = validateRedirect(req.query?.redirect || '/');
+      passport.authenticate(
+        'local',
+        authenticate_return(req, res, next, redirect)
+      )(req, res, next);
+    }
+  );
+
+  /**
+   * Register for a notebook using an invite, if no existing account
+   * then ask them to register.  User is authenticated in either case.
+   * Return a redirect response to the given URL
+   */
+  app.get<{invite_id: string}, {}, {}, RequestQueryRedirect>(
+    '/register/:invite_id/',
+    async (req, res) => {
+      const redirect = validateRedirect(req.query?.redirect || '/');
+      const invite_id = req.params.invite_id;
+      req.session['invite'] = invite_id;
+      const invite = await getInvite(invite_id);
+      if (!invite) {
+        res.render('invite-error', {redirect});
+      } else if (req.user) {
+        // user already registered, sign them up for this notebook
+        // should there be conditions on this? Eg. check the email.
+        await acceptInvite(req.user, invite);
+        redirect_with_token(res, req.user, redirect);
+      } else {
+        // need to sign up the user, show the registration page
+        const available_provider_info = [];
+        for (const handler of CONDUCTOR_AUTH_PROVIDERS) {
+          available_provider_info.push({
+            label: handler,
+            name: AVAILABLE_AUTH_PROVIDER_DISPLAY_INFO[handler].name,
+          });
+        }
+        const encodedRedirect = encodeURIComponent(
+          `/register/${invite_id}?redirect=${redirect}`
+        );
+        res.render('register', {
+          invite: invite_id,
+          loginURL: `/auth?redirect=${encodedRedirect}`,
+          providers: available_provider_info,
+          redirect: redirect,
+          localAuth: true, // maybe make this configurable?
+          messages: req.flash(),
         });
       }
```

Comment:
Just checking that it's possible to login from the register page. The workflow here would be 1) open app 2) app is logged out for some reason 3) accept code 4) redirected to register, but user already has account.